File: Contest/20180325-hihocoder-52/B.cpp
#include <cstdio>
#include <cstring>
#include <iostream>
#include <algorithm>
#include <map>

using namespace std;

const int dx[4] = {0, 1, 0, -1};
const int dy[4] = {1, 0, -1, 0};
map<string, long long> dp;
string M[25];
int n, m;

bool isOK(int x, int y){
	for(int i = 0; i < 4; i++){
		int nx = x + dx[i];
		int ny = y + dy[i];
		if(nx < 0 || nx >= n || ny < 0 || ny >= m)
		  continue;
		if(M[nx][ny] == '1')return true;
	}
	return false;
}

string toString(){
	string str(n*m, '0');
	for(int i = 0; i < n; i++)
	  for(int j = 0; j < m; j++)
		str[i*m+j] = M[i][j];
	return str;
}

long long dfs(int mie){
	if(mie == 1)return 1;
	if(dp[toString()] != 0)return dp[toString()];
	dp[toString()] = 0;
	long long sum = 0;
	for(int i = 0; i < n; i++){
		for(int j = 0; j < m; j++){
			if(M[i][j] == '0' && isOK(i, j)){
				M[i][j] = '1';
				sum += dfs(mie-1);
				M[i][j] = '0';
			}
		}
	}
	return dp[toString()] = sum;
}

int main()
{
	while(cin>>n>>m){
		int mie = 0;
		dp.clear();
		for(int i = 0; i < n; i++){
			cin>>M[i];
			for(int j = 0; j < m; j++){
				if(M[i][j] == '0')mie++;
			}
		}
		cout<<dfs(mie)<<endl;
	}

	return 0;
}

File: Contest/20180325-hihocoder-52/B.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Main {
	static int[] dx = {0, 1, 0, -1};
	static int[] dy = {1, 0, -1, 0};
	static StringBuilder[] M = new StringBuilder[20];
	static int n, m;
	static Map<String, Long> dp = new HashMap<String, Long>();
	
	static String MtoString() {
		StringBuilder sb = new StringBuilder("");
		for(int i = 0; i < n; i++) {
			sb.append(M[i]);
		}
		return sb.toString();
	}
	
	static boolean isOK(int x, int y) {
		for(int i = 0; i < 4; i++){
	        int nx = x + dx[i];
	        int ny = y + dy[i];
	        if(nx < 0 || nx >= n || ny < 0 || ny >= m)
	          continue;
	        if(M[nx].charAt(ny) == '1')return true;
	    }
	    return false;
	}
	
	static long dfs(int mie) {
		if(mie == 1)return 1;
		if(dp.get(MtoString()) != null)return dp.get(MtoString());
		long sum = 0;
		for(int i = 0; i < n; i++){
	        for(int j = 0; j < m; j++){
	            if(M[i].charAt(j) == '0' && isOK(i, j)){
	                M[i].setCharAt(j, '1');
	                sum += dfs(mie-1);
	                M[i].setCharAt(j, '0');
	            }
	        }
	    }
		dp.put(MtoString(), sum);
		return sum;
	}
	
	public static void main(String[] args) {
		Scanner cin = new Scanner(System.in);
		
		int mie;
		while(cin.hasNext()) {
			dp.clear();
			n = cin.nextInt();
			m = cin.nextInt();
			mie = 0;
			for(int i = 0; i < n; i++) {
				M[i] = new StringBuilder(cin.next());
				for(int j = 0; j < m; j++)
					if(M[i].charAt(j) == '0')
						mie++;
			}
			System.out.println(dfs(mie));
		}
	}
}
